### PR TITLE
azurerm_web_application_firewall_policy: add http_listener_ids and path_based_rule_ids

### LINKED ISSUE #10859: Support for HTTP Listener and Path Based Rules in Web Application Firewall Policy
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

There are three ways to associate a WAF Policy with an Application Gateway:

1. The Application Gateway can be set to use the WAF Policy as its global policy
2. The WAF Policy can be configured to be associated with one or more of the Application Gateway's HTTP Listeners
3. The WAF Policy can be configured to be associated with one or more of the Application Gateway's Path Based Rules

### New or Affected Resource(s)

* azurerm_application_gateway
* azurerm_web_application_firewall_policy

### Potential Terraform Configuration

Based on test configuration:

```hcl
provider "azurerm" {
  features {}
}

resource "azurerm_resource_group" "test" {
  name     = "acctestRG-%d"
  location = "%s"
}

resource "azurerm_virtual_network" "test" {
  name                = "acctest-vnet-%d"
  resource_group_name = azurerm_resource_group.test.name
  address_space       = ["10.0.0.0/16"]
  location            = azurerm_resource_group.test.location
}

resource "azurerm_subnet" "test" {
  name                 = "subnet-%d"
  resource_group_name  = azurerm_resource_group.test.name
  virtual_network_name = azurerm_virtual_network.test.name
  address_prefix       = "10.0.0.0/24"
}

resource "azurerm_public_ip" "test" {
  name                = "acctest-pubip-%d"
  location            = azurerm_resource_group.test.location
  resource_group_name = azurerm_resource_group.test.name
  allocation_method   = "Dynamic"
}

# since these variables are re-used - a locals block makes this more maintainable
locals {
  backend_address_pool_name      = "${azurerm_virtual_network.test.name}-beap"
  frontend_port_name             = "${azurerm_virtual_network.test.name}-feport"
  frontend_ip_configuration_name = "${azurerm_virtual_network.test.name}-feip"
  http_setting_name              = "${azurerm_virtual_network.test.name}-be-htst"
  listener_name                  = "${azurerm_virtual_network.test.name}-httplstn"
  request_routing_rule_name      = "${azurerm_virtual_network.test.name}-rqrt"
}

resource "azurerm_public_ip" "teststd" {
  name                = "acctest-PubIpStd-%[2]d"
  location            = azurerm_resource_group.test.location
  resource_group_name = azurerm_resource_group.test.name
  allocation_method   = "Static"
  sku                 = "Standard"
}

resource "azurerm_application_gateway" "test" {
  name                = "acctestag-%[2]d"
  resource_group_name = azurerm_resource_group.test.name
  location            = azurerm_resource_group.test.location

  sku {
    name     = "WAF_v2"
    tier     = "WAF_v2"
    capacity = 2
  }

  gateway_ip_configuration {
    name      = "my-gateway-ip-configuration"
    subnet_id = azurerm_subnet.test.id
  }

  frontend_port {
    name = local.frontend_port_name
    port = 80
  }

  frontend_ip_configuration {
    name                 = local.frontend_ip_configuration_name
    public_ip_address_id = azurerm_public_ip.teststd.id
  }

  backend_address_pool {
    name = local.backend_address_pool_name
  }

  backend_http_settings {
    name                  = local.http_setting_name
    cookie_based_affinity = "Disabled"
    port                  = 443
    protocol              = "Https"
    request_timeout       = 1

    pick_host_name_from_backend_address = true
  }

  http_listener {
    name                           = local.listener_name
    frontend_ip_configuration_name = local.frontend_ip_configuration_name
    frontend_port_name             = local.frontend_port_name
    protocol                       = "Http"
    firewall_policy_id             = azurerm_web_application_firewall_policy.testfwp_listener.id
  }

  request_routing_rule {
    name                       = local.request_routing_rule_name
    rule_type                  = "Basic"
    http_listener_name         = local.listener_name
    backend_address_pool_name  = local.backend_address_pool_name
    backend_http_settings_name = local.http_setting_name
  }

  url_path_map {
    name                               = local.url_path_map_name
    default_backend_address_pool_name  = local.backend_address_pool_name
    default_backend_http_settings_name = local.http_setting_name

    path_rule {
      name                       = local.path_rule_name
      backend_address_pool_name  = local.backend_address_pool_name
      backend_http_settings_name = local.http_setting_name

      paths = [
        "/test",
      ]
    }
  }
}

resource "azurerm_web_application_firewall_policy" "testfwp" {
  name                = "acctest-fwp-%[2]d"
  resource_group_name = azurerm_resource_group.test.name
  location            = azurerm_resource_group.test.location

  policy_settings {
    enabled = true
    mode    = "Prevention"
  }

  managed_rules {
    managed_rule_set {
      type    = "OWASP"
      version = "3.1"
    }
  }
}

resource "azurerm_web_application_firewall_policy" "testfwp_listener" {
  name                = "acctest-fwp-listener-%[2]d"
  resource_group_name = azurerm_resource_group.test.name
  location            = azurerm_resource_group.test.location
  http_listener_ids   = [for v in azurerm_application_gateway.test.http_listener : v.id]
  path_based_rule_ids = flatten([for v in azurerm_application_gateway.test.url_path_map : [for w in v.path_rule : w.id]])

  policy_settings {
    enabled = true
    mode    = "Prevention"
  }

  managed_rules {
    managed_rule_set {
      type    = "OWASP"
      version = "3.1"
    }
  }
}
```

### References

N/A

### PR DESCRIPTION
**Pull request by sirlatrom**
Fixes #10859.

Not sure about the syntax in the acceptance test when referencing the path based rule IDs.

### CODE REVIEW

**Comment by sirlatrom** [comment]
@favoretti I'd like your input on whether I've placed the subresource ID parse/validate functions in the right files, as my first commit placed them inside files that are overwritten by go:generate.

**Comment by favoretti** [comment]
@sirlatrom I need to take a deeper look at this, but can't you just use the generators to generate these ID functions, rather than creating them by hand?

**Comment by sirlatrom** [comment]
> @sirlatrom I need to take a deeper look at this, but can't you just use the generators to generate these ID functions, rather than creating them by hand?

@favoretti I'd love to! How do I get started?

**Comment by favoretti** [comment]
@sirlatrom Just add them [here](https://github.com/terraform-providers/terraform-provider-azurerm/blob/master/azurerm/internal/services/network/resourceids.go) and go generate will generate them for you.

File: azurerm/internal/services/network/web_application_firewall_policy_resource.go [hunk 1]
```diff
@@ -282,6 +282,26 @@ func resourceWebApplicationFirewallPolicy() *schema.Resource {
...
+				},
+			},
+
+			"path_based_rule_ids": {
+				Type:     schema.TypeList,
+				Optional: true,
+				Elem: &schema.Schema{
+					Type:         schema.TypeString,
+					Required:     true,
+					ValidateFunc: validate.ApplicationGatewayPathBasedRuleID,
```
Comment (favoretti) [inline azurerm/internal/services/network/web_application_firewall_policy_resource.go:301]: I assume this wants to become `ApplicationGatewayURLPathMapPathRule`.

Comment (sirlatrom) [reply azurerm/internal/services/network/web_application_firewall_policy_resource.go]: Thanks, I was a little too quick on the trigger there 😅

**Comment by sirlatrom** [comment]
Rebased on origin/master.

**Comment by sirlatrom** [comment]
@katbyte Is there any chance this PR could get on the next milestone?

File: azurerm/internal/services/network/web_application_firewall_policy_resource.go [hunk 2]
```diff
@@ -376,8 +400,11 @@ func resourceWebApplicationFirewallPolicyRead(d *schema.ResourceData, meta inter
 		if err := d.Set("managed_rules", flattenWebApplicationFirewallPolicyManagedRulesDefinition(webApplicationFirewallPolicyPropertiesFormat.ManagedRules)); err != nil {
 			return fmt.Errorf("Error setting `managed_rules`: %+v", err)
 		}
-		if err := d.Set("managed_rules", flattenWebApplicationFirewallPolicyManagedRulesDefinition(webApplicationFirewallPolicyPropertiesFormat.ManagedRules)); err != nil {
-			return fmt.Errorf("Error setting `managed_rules`: %+v", err)
```
Comment (katbyte) [inline azurerm/internal/services/network/web_application_firewall_policy_resource.go:380]: Is there a reason we are removing managed rules from here?

**Review by katbyte — CHANGES_REQUESTED** [review]
Thanks for the pr @sirlatrom - aside from one comment i've left inline this is looking good!

File: azurerm/internal/services/network/web_application_firewall_policy_resource.go (hunk 2 above)
Comment (sirlatrom) [reply azurerm/internal/services/network/web_application_firewall_policy_resource.go:380]: It's a good question, but the answer is above, as the lines were duplicated by mistake in my original contribution of `managed_rules`:

https://github.com/terraform-providers/terraform-provider-azurerm/commit/1b5a0b36d5c3a756e2efe45a3eff2bda67135997#diff-25ef614a1fbe2de62ec91d10ddde72c778c5876a77dc676908b065a8268846c5R344-R349

So it _should_ be safe 😅

Comment (katbyte) [reply azurerm/internal/services/network/web_application_firewall_policy_resource.go:380]: 🤦‍♀️  - that they are and i'membarrassedd that i missed that

**Review by katbyte — APPROVED** [review]
LGTM 👍

File: azurerm/internal/services/network/web_application_firewall_policy_resource.go (hunk 2 above)
Comment (sirlatrom) [reply azurerm/internal/services/network/web_application_firewall_policy_resource.go:380]: > 🤦‍♀️  - that they are and i'membarrassedd that i missed that

Don't be! I bet you have a lot on your plate with all the things happening in this provider. Thanks for the review! 😊

**Pull request merged**